### PR TITLE
core(lhr): make reportCategories shallow; move audit scores to AuditResultJSON

### LINKED ISSUE #1999: Reduce size of JSON results: remove duplication
We currently duplicate a lot of content in `audits` and `reportCategories[x].audits.result`. `extendedInfo` is going away, but `details`, meta data, and other audit info is duplicated in both. sections. For a complex site, this can make our JSON results huge (e.g. many event listeners, urls, etc.). If the audit is used in multiple categories, the problem only gets worse :(

Example:

```
"audits": {
    "is-on-https": {
      "score": false,
      "displayValue": "1 insecure request found",
      "rawValue": false,
      "extendedInfo": {
        "formatter": "url-list",
        "value": [
          {
            "url": "ajax.googleapis.com/…3.2.1/jquery.min.js"
          }
        ]
      },
      "scoringMode": "binary",
      "name": "is-on-https",
      "category": "Security",
      "description": "Uses HTTPS",
      "helpText": "All sites should be protected with HTTPS, even ones that don't handle sensitive data. HTTPS prevents intruders from tampering with or passively listening in on the communications between your app and your users, and is a prerequisite for HTTP/2 and many new web platform APIs. [Learn more](https://developers.google.com/web/tools/lighthouse/audits/https).",
      "details": {
        "type": "list",
        "header": {
          "type": "text",
          "text": "Insecure URLs:"
        },
        "items": [
          {
            "type": "text",
            "text": "ajax.googleapis.com/…3.2.1/jquery.min.js"
          }
        ]
      }
    },
...
"reportCategories": [
    {
      "name": "Progressive Web App",
      "weight": 1,
      "description": "These audits validate the aspects of a Progressive Web App. They are a subset of the [PWA Checklist](https://developers.google.com/web/progressive-web-apps/checklist).",
      "audits": [
        {
          "id": "is-on-https",
          "weight": 1,
          "result": {
            "score": false,
            "displayValue": "1 insecure request found",
            "rawValue": false,
            "extendedInfo": {
              "formatter": "url-list",
              "value": [
                {
                  "url": "ajax.googleapis.com/…3.2.1/jquery.min.js"
                }
              ]
            },
            "scoringMode": "binary",
            "name": "is-on-https",
            "category": "Security",
            "description": "Uses HTTPS",
            "helpText": "All sites should be protected with HTTPS, even ones that don't handle sensitive data. HTTPS prevents intruders from tampering with or passively listening in on the communications between your app and your users, and is a prerequisite for HTTP/2 and many new web platform APIs. [Learn more](https://developers.google.com/web/tools/lighthouse/audits/https).",
            "details": {
              "type": "list",
              "header": {
                "type": "text",
                "text": "Insecure URLs:"
              },
              "items": [
                {
                  "type": "text",
                  "text": "ajax.googleapis.com/…3.2.1/jquery.min.js"
                }
              ]
            }
          },
          "score": 0
        },
```

### PR DESCRIPTION
**Pull request by paulirish**
### Changes

* `reportCategories` no longer contains real auditResults (aka _de-dupe all audit results_ from #4333)
   * The categories still keep their `.score` prop. Eg: `reportCategories.performance.score`
  * Each audit's score moves from `reportCategories.performance['audit-name']` (which now just holds `weight` and `group`) over to `audits['audit-name']`
* Remove `LHR.score`, the overallScore

---------------

This is a chain of PRs. Landing order is... 1st: `newdetails` (#4616). 2nd: `shallowCategories` (#4711). 3rd: `scoring2.0` (#4690)

ref #4614

fixes #1999

### CODE REVIEW

File: lighthouse-core/runner.js [hunk 1]
```diff
@@ -93,9 +93,11 @@ class Runner {
       const resultsById = {};
       for (const audit of runResults.auditResults) resultsById[audit.name] = audit;
 
-      let report;
+      let categories;
       if (opts.config.categories) {
-        report = Runner._scoreAndCategorize(opts, resultsById);
+        Runner._scoreAndCategorize(opts, resultsById);
+        // COMPAT: after dropping Node 6, upgrade to use Object.values
```
Comment (patrickhulce) [inline lighthouse-core/runner.js]: does this mean we can do it now? :)

File: lighthouse-core/scoring.js [hunk 2]
```diff
@@ -13,30 +13,33 @@ class ReportScoring {
...
   static scoreAllCategories(config, resultsByAuditId) {
-    const categories = Object.keys(config.categories).map(categoryId => {
+    Object.keys(config.categories).forEach(categoryId => {
       const category = config.categories[categoryId];
       category.id = categoryId;
 
-      const audits = category.audits.map(audit => {
+      category.audits.forEach(audit => {
         const result = resultsByAuditId[audit.id];
         // Cast to number to catch `null` and undefined when audits error
```
Comment (patrickhulce) [inline lighthouse-core/scoring.js]: should all of these line just be moved into `audit.js` then to keep all the score mapping in the same place?

Comment (patrickhulce) [reply lighthouse-core/scoring.js]: more a comment for #4690 I suppose

**Comment by brendankenny** [comment]
smokehouse failures?

File: lighthouse-core/runner.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/runner.js]: do it!

File: lighthouse-core/scoring.js (hunk 2 above)
Comment (paulirish) [reply lighthouse-core/scoring.js]: Aye. It actually already happens in that PR: https://github.com/GoogleChrome/lighthouse/pull/4690/files#diff-488f5da33160a651526463bdb951e5deL31

**Comment by paulirish** [comment]
> As discussed, we'll want to smoosh the audit results into a copy of the LHR so the canonical one still exists untouched.

Done. added a test.

> something for the future: as we add more typechecking, we may want to split AuditJSON to a version with result (only used inside the report renderer) and a version without (used everywhere else) so that we can more easily reason about which we have at some point in time and so we don’t have to put in a bunch of checks to keep tsc happy :)

YESSSSSSS. So want this.

Here was my typechecking spike, validating the outgoing LHR against some real types. Works! https://github.com/GoogleChrome/lighthouse/compare/95e19a56503c702a5077c62dda1112cd39c4d705...typechecking


---------

This and #4690 are updated and ready

File: lighthouse-core/report/v2/renderer/report-renderer.js [hunk 3]
```diff
@@ -155,24 +157,39 @@ class ReportRenderer {
     perfCategoryRenderer.setTemplateContext(this._templateContext);
 
     const categories = reportSection.appendChild(this._dom.createElement('div', 'lh-categories'));
+
+    ReportRenderer.smooshAuditResultsIntoCategories(report.audits, report.reportCategories);
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/report-renderer.js]: does this make more sense up in `renderReport` before passing into `_renderReport`? Then internally it'll always be the smooshed version and we won't have to remember not to use `reportCategories` above this line

File: lighthouse-core/report/v2/renderer/report-renderer.js [hunk 4]
```diff
@@ -181,6 +198,16 @@ if (typeof module !== 'undefined' && module.exports) {
...
 
+/**
+ * @typedef {{
+ *     id: string,
+ *     weight: number,
+ *     group: (string|undefined),
+ *     result: (ReportRenderer.AuditResultJSON|undefined)
+ * }}
+ */
+ReportRenderer.AuditJSON; // eslint-disable-line no-unused-expressions
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/report-renderer.js]: defined below (merge conflict, I think)

File: lighthouse-core/runner.js [hunk 5]
```diff
@@ -108,8 +109,7 @@ class Runner {
         audits: resultsById,
         artifacts: runResults.artifacts,
         runtimeConfig: Runner.getRuntimeConfig(opts.flags),
-        score: report ? report.score : 0,
-        reportCategories: report ? report.categories : [],
+        reportCategories: categories,
```
Comment (brendankenny) [inline lighthouse-core/runner.js]: before this defaulted to an empty array. Will it cause problems anywhere as `undefined` (and removed from the JSON) instead?

File: lighthouse-core/scoring.js [hunk 6]
```diff
@@ -13,35 +13,49 @@ class ReportScoring {
...
    * Returns the report JSON object with computed scores.
-   * @param {{categories: !Object<string, {id: string|undefined, weight: number|undefined, audits: !Array<{id: string, weight: number|undefined}>}>}} config
+   * @param {{categories: !Object<string, {id: string|undefined, weight: number|undefined, score: number|undefined, audits: !Array<{id: string, weight: number|undefined}>}>}} config
    * @param {!Object<string, {score: ?number|boolean|undefined, notApplicable: boolean, informative: boolean}>} resultsByAuditId
    * @return {{score: number, categories: !Array<{audits: !Array<{score: number, result: !Object}>}>}}
    */
   static scoreAllCategories(config, resultsByAuditId) {
-    const categories = Object.keys(config.categories).map(categoryId => {
-      const category = config.categories[categoryId];
+    for (const [categoryId, category] of Object.entries(config.categories)) {
```
Comment (brendankenny) [inline lighthouse-core/scoring.js]: 🎉

**Comment by paulirish** [comment]
merge when green.

**Pull request merged**